Reject policies that contain rules with more than one L3 match in a single rule

joestringer opened the pull request:
This PR rejects policies which are not supported as per the assessment in #3004 .

The combination of `fromCIDR` and `fromEndpoints` was previously allowed, but conflicted with the standard rules for members in an `IngressRule` defined below, so this PR disallows it.

https://github.com/cilium/cilium/blob/ca6e1cba72621a0911b93ebfc0e94b5b47f7ce45/pkg/policy/api/rule.go#L99-L102

Note that there are certain combinations, like the above, which we have previously accepted policies for and treated as OR (eg, `FromCIDR: x` with `FromEndpoints: y` in the same rule). These policies must be split into separate rules, one for each L3 after this PR. Examples can be seen in the unit test changes in this PR.

joestringer committed:
policy: Reject all unsupported rule combinations

Signed-off-by: Joe Stringer <joe@covalent.io>

manalibhutiyani commented on pkg/policy/api/rule_validation.go:
```diff
@@ -81,8 +98,29 @@ func (i *IngressRule) sanitize() error {
+		}
+	}
+	for member := range l3Members {
+		if l3Members[member] > 0 && len(e.ToPorts) > 0 && !l3DependentL4Support[member] {
```
Do we still need the `!l3DependentL4Support[member]` check in egress, since all values are false?

joestringer replied:
Not strictly, no, but egress label-dependent l4 is pretty close so it seems easy enough to just include this here. It'll require a minor rebase and update for the egress work.

rlenglet reviewed (approved):
I like this. Thanks!

joestringer committed:
test: Fix formatting in headless policy

Signed-off-by: Joe Stringer <joe@covalent.io>

nebril commented on pkg/policy/api/rule_validation.go:
```diff
@@ -81,8 +98,29 @@ func (i *IngressRule) sanitize() error {
+	}
+	for m1 := range l3Members {
+		for m2 := range l3Members {
+			if m2 != m1 && l3Members[m1] > 0 && l3Members[m2] > 0 {
```
This will cause ToServices to break, as ToServices rule is being translated into ToCIDRSet rule. This code needs a bit more logic that handles that case.

joestringer replied:
Cool. Rather than opening a hole for configuration of things that violate the general rule semantics, I've attempted to reshuffle the sanitization logic so that internally we can represent things in whichever way makes sense, but externally we should only advertise support for particular combinations. This approach is represented by the new patch "daemon: Shift policy sanitize up to handle".

joestringer committed:
daemon: Shift policy sanitize up to handle

Rules should be sanitized relatively early, and in the k8s case they
already were sanitized before reaching the repository. Do likewise for
the regular API case so that k8s doesn't double-sanitize the rules.

This avoids issues when performing translation of k8s rules internally
after we've already sanitized and validated rule combinations, where the
translated rules no longer satisfy the same constraints as what is
rejected at the API - for example, translating headless services policy
into CIDR policies.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Reuse repo.AddList() from repo.Add()

Share more code between these functions.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf reviewed (approved):
I'm not 100% happy with the validation only happening via the API but this is a great step forward so I'm fine with this for now.
